Add streaming subprotocol negotiation

ncdc opened the pull request:
Add streaming subprotocol negotiation for exec, attach, and port
forwarding. Restore previous (buggy) exec functionality as an
unspecified/unversioned subprotocol so newer kubectl clients can work
against 1.0.x kubelets.

Fixes #15702
Fixes #15726

Commit:
Add streaming subprotocol negotiation

Add streaming subprotocol negotiation for exec, attach, and port
forwarding. Restore previous (buggy) exec functionality as an
unspecified/unversioned subprotocol so newer kubectl clients can work
against 1.0.x kubelets.

ncdc commented:
This is my WIP for getting this implemented. I still need to flesh out some unit tests, and I'd like to restructure some of the exec/attach/portforward server code to make it easier to add new subprotocol version implementations, but that may have to wait depending on timing.

@smarterclayton @lavalamp @liggitt @kubernetes/rh-cluster-infra @bgrant0607

Commit:
Move protocol handlers to separate files

lavalamp commented:
Made a quick skim; nothing stands out. Thanks for doing this!

bronger commented:
The following is tested against current GKE.  The command

    kubectl exec -ti single_container_pod bash

quits without taking input with kubectl 1.1, however, with this PR, I can happily use the bash interactively.

Something like

    echo "Hello" | kubectl_old exec -i postgresql-vahuk 'bash -c "cat > /tmp/toll"'

hangs as with kubectl 1.0, but if I understand the PR correctly, this is hangxing as expected, isn't it?

ncdc commented:
@bronger yes. This PR restores the previous 1.0 behavior when a newer kubectl connects to a 1.0.x kubelet. There is no way to eliminate that hang unless we backport the fix to 1.0.x. The hang is fixed with new kubectl + new kubelet.

lavalamp commented:
@ncdc What needs to be done still before we can review & merge?

ncdc commented:
@lavalamp I've been looking at refactoring exec & attach so the code is better able to support multiple protocol handlers internally, but I don't think that's necessary to fix this issue. I have 1 local commit that I haven't pushed yet that does the same type of refactoring for port forward, and I think that's ready to go assuming the units pass (I need to double-check). I'll push as soon as that's ready, and then we can move forward with this and defer the exec/attach changes to later, if that sounds good?

ncdc commented:
@lavalamp updated, PTAL

lavalamp commented:
@smarterclayton do you have time to review this or would you like me to take it?

ncdc commented:
If it would make it simpler, I could back out the 3rd commit (creating a port forward protocol handler) as it's not strictly necessary to fix the original issue (and then bring it back in a follow-up).

smarterclayton commented:
Let's split the third commit into another pull to streamline this.

ncdc commented:
Ok, I'll save it off to the side, rebase to strip that out, and then start addressing your comments.

ncdc commented:
I've got to step away for a couple of hours, but I will resume later this evening.

Commit:
Address code review comments

ncdc commented:
Trying to figure out why doc verification is failing...

ncdc commented:
Figured out docs issue. Fixing.

Commit:
Move port forward protocol constant to subpackage

Move port forward protocol name constant to a subpackage underneath
pkg/kubelet to avoid flags applicable to the kubelet leaking into
kubectl. Eventually, handlers for specific protocol versions will move
into the new subpackage as well.

ncdc commented:
To avoid an import cycle, I had placed a new constant in pkg/kubelet/server.go to be used by both the client and server port forwarding code. Unfortunately, this had the side effect of pulling in all kubelet-related flags into kubectl. I have moved the constant into a new subpackage under pkg/kubelet so the extra flags aren't pulled in. Eventually I will be moving protocol handlers and ServePortForward into this package as well.

ncdc commented:
@smarterclayton ready for another round of review

ncdc commented:
e2e flake... `[Fail] Pods [It] should not back-off restarting a container on LivenessProbe failure `

@k8s-bot test this

smarterclayton commented:
What e2e test cover this?  Can we call the old protocol and verify it behaves a certain way from an e2e test somehow?

ncdc commented:
We have e2e tests for exec:

- no stdin
- non-interactive stdin
- pseudo-interactive stdin

These all use the kubectl built with the release. There is no way currently to specify the subprotocol you want. The only way right now to test that is to run the kubectl from this branch against a 1.0.x kubelet...

ncdc commented:
Additionally, for the pseudo-interactive stdin test case, as mentioned in #15726, it passes with a kubectl from master and a 1.0 kubelet, even though when you manually run the same test, it doesn't work. It appears to be a timing issue in the test itself, so I'm not sure of any reliable way to test that.

smarterclayton commented:
Can you open an issue to have an integration test that mocks a backend that
verifies the client logic is correct?  We don't need it in 1.1 (kube or
openshift) but we will want it.

On Thu, Oct 22, 2015 at 3:36 PM, Andy Goldstein <notifications@github.com>
wrote:

> Additionally, for the pseudo-interactive stdin test case, as mentioned in
> #15726 <https://github.com/kubernetes/kubernetes/issues/15726>, it passes
> with a kubectl from master and a 1.0 kubelet, even though when you manually
> run the same test, it doesn't work. It appears to be a timing issue in the
> test itself, so I'm not sure of any reliable way to test that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/15961#issuecomment-150332588>
> .
>

ncdc commented:
Issue opened: #16119

smarterclayton commented on pkg/util/httpstream/httpstream.go:
```diff
@@ -96,3 +102,44 @@ func IsUpgradeRequest(req *http.Request) bool {
+// indicating the chosen subprotocol. If no match is found, HTTP forbidden is
+// returned, along with a response header containing the list of protocols the
+// server can accept.
+func Handshake(req *http.Request, w http.ResponseWriter, serverProtocols []string, defaultProtocol string) (string, error) {
```
Unit test for this

smarterclayton commented on pkg/client/unversioned/remotecommand/remotecommand.go:
```diff
@@ -97,155 +95,98 @@ func NewStreamExecutor(upgrader httpstream.UpgradeRoundTripper, fn func(http.Rou
+
+// Stream opens a protocol streamer to the server and streams until a client closes
+// the connection or the server disconnects.
+func (e *streamExecutor) Stream(stdin io.Reader, stdout, stderr io.Writer, tty bool) error {
```
Unit test for this

Commit:
Add httpstream.Handshake unit test

ncdc commented:
@smarterclayton unit test added for `httpstream.Handshake`. You also asked for a test for `Stream`... that's already partially covered by `TestRequestExecuteRemoteCommand`, although the only thing it tests w.r.t protocol negotiation is that the client & server are both using v2.channel.k8s.io. If you want a unit test that the client selects the correct `streamProtocolHandler`, I think `Stream` will need to be refactored a bit.

smarterclayton commented:
This is LGTM for now.

lavalamp commented:
@ncdc did/can you cherry-pick this?

ncdc commented:
@lavalamp I have not cherry picked it, but we do need to pull it into 1.1. I won't have time to do it tonight, but can do it tomorrow morning if needed.

lavalamp commented:
@ncdc Yes, please do, thanks!

On Mon, Oct 26, 2015 at 3:53 PM, Andy Goldstein <notifications@github.com>
wrote:

> @lavalamp <https://github.com/lavalamp> I have not cherry picked it, but
> we do need to pull it into 1.1. I won't have time to do it tonight, but can
> do it tomorrow morning if needed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/15961#issuecomment-151307528>
> .
>

ncdc commented:
@lavalamp @smarterclayton in order to cherry pick, it looks like we'll need pull in #12283 (port forward backwards compat fixes) and #14618 (exec refactoring to use it without a client.Config) first. Is that ok?

lavalamp commented:
@ncdc Yes, I think that's OK, thanks.
